osquery/tables/system/windows/startup_items.cpp

```diff
@@ -60,34 +62,55 @@ const std::set<std::string> kStartupStatusRegKeys = {
 // Starts with 0[0-9] but not followed by all 0s
 const auto kStartupDisabledRegex = boost::regex("^0[0-9](?!0+$).*$");
 
-static inline void parseStartupPath(const std::string& path, Row& r) {
-  std::string expandedPath = path;
+bool parseStartupPath(const std::string& entry, Row& r) {
+  // In case the path is already quoted, splitArgs(...) extract correctly the
+  // path with/without spaces
+  const auto argsp = splitArgs(entry);
 
-  // NOTE(ww): This is a pretty dumb expansion test, but the query that feeds
-  // us these paths doesn't pass us REG_EXPAND_SZ or any other hint
-  // that we could use instead.
-  if (path.find('%') != std::string::npos) {
-    if (const auto expanded = expandEnvString(path)) {
-      expandedPath = *expanded;
-    }
+  if (!argsp.has_value()) {
+    return false;
   }
 
-  if (pathExists(fs::path(expandedPath)).ok()) {
-    r["path"] = expandedPath;
-  } else {
-    if (const auto argsp = splitArgs(expandedPath)) {
-      auto args = *argsp;
+  if (pathExists(entry)) {
+    r["path"] = entry;
+    return true;
+  }
 
-      r["path"] = args[0];
+  auto args = argsp.value();
 
-      if (args.size() > 1) {
-        args.erase(args.begin());
-        r["args"] = boost::join(args, " ");
-      }
-    } else {
-      r["path"] = expandedPath;
+  // We already tested for emptyness
+  auto path = args[0];
+
+  if (args.size() == 1) {
+    r["path"] = path;
+    return true;
+  }
+
+  if (pathExists(path)) {
+    r["path"] = path;
+    r["args"] = boost::join(
+        std::vector<std::string>{std::begin(args) + 1, std::end(args)}, " ");
+    return true;
+  }
+
+  auto end = std::end(args);
+
+  std::vector<std::string> tmp_path(std::begin(args), end);
+
+  while (!pathExists(path)) {
+    path = boost::join(tmp_path, " ");
+    tmp_path = std::vector<std::string>{std::begin(args), --end};
+
+    if (std::begin(args) == end) {
+      return false;
```

Comment:
I can see why you decided to do this, but it does have the side effect of returning an empty string for `path` if no valid path is detected. I wonder if it's better to just return the entire entry in this case?

Reply:
If the `path` is detected as invalid we skip producing a row entirely. Do you mean you prefer having the Row even if the path is detected as invalid or do you mean something else ? My reasoning was that an invalid path should not produce any row, but I understand that we might actually want to have a Row for invalid paths because it could be a concern.

Reply:
Yes I misunderstood but I do think an incomplete/invalid entry still makes sense to return -- since those values are actually configured on the system, osquery should return them (IMO).

Reply:
I have just remembered why I decided to not return an empty row: we cannot guarantee the path in the entry is double quoted when it includes spaces, so we are always going to split it into multiple components (the bug we are trying to fix). i.e. for the entry `C:\Program Files\some space\file.exe` is going to be split into `C:\Program` and `Files\some` and `space\file.exe`.  And as far as I know it is impossible to reliably "understand" when we have a path unless the file exists on the filesystem, which is not the case when the entry is invalid/incomplete as the case we are trying to fix here. It is a catch 22.

Hence the decision of not including the entry in the result set.

Reply:
Huh... That behavior sounds horrific.